src/cache/in_memory_cache.py

```diff
@@ -0,0 +1,107 @@
+from collections import defaultdict, OrderedDict
+from typing import Union
+import threading
+import src.constants as constants
+
+
+class ListNode:
+    """A node class for the doubly linked list used in the InMemoryCache."""
+
+    def __init__(self, key, value):
+        """Initialize a ListNode with a key, value, and frequency."""
+        self.key = key
+        self.val = value
+        self.freq = 1
+
+
+class InMemoryCache:
+    """An in-memory LFU cache implementation using a combination of a dictionary and an OrderedDict."""
```

Comment:
you've converted this from an LRU to an LFU cache, but that is not what we want for this use case.

If a user iterates on a single prompt 100 times, and then goes away, and then another 50 users each iterate twice, the first cache entry we want to throw away is the old user who iterated 100 times but hasn't been back recently, not the 50 users who are currently active.

What matters is how recently the cache entry was referenced, not how many times.

Reply:
Changed it to LRU.